os/net/app-layer/snmp/snmp-ber.c

```diff
@@ -0,0 +1,274 @@
+  for(i = 0; i < length; ++i) {
+    (*out_len)++;
+    *out-- = (uint8_t)*str--;
+  }
+
+  out = snmp_ber_encode_length(out, out_len, length);
+  out = snmp_ber_encode_type(out, out_len, BER_DATA_TYPE_OCTET_STRING);
+
+  return out;
+}
+/*---------------------------------------------------------------------------*/
+unsigned char *
+snmp_ber_encode_null(unsigned char *out, uint32_t *out_len, uint8_t type)
+{
+  (*out_len)++;
+  *out-- = 0x00;
+  out = snmp_ber_encode_type(out, out_len, type);
+
+  return out;
+}
+/*---------------------------------------------------------------------------*/
+unsigned char *
+snmp_ber_decode_type(unsigned char *buff, uint32_t *buff_len, uint8_t *type)
+{
+  *type = *buff++;
+  (*buff_len)--;
+
+  return buff;
+}
+/*---------------------------------------------------------------------------*/
+unsigned char *
+snmp_ber_decode_length(unsigned char *buff, uint32_t *buff_len, uint8_t *length)
+{
+  *length = *buff++;
+  (*buff_len)--;
+
+  return buff;
+}
+/*---------------------------------------------------------------------------*/
+unsigned char *
+snmp_ber_decode_integer(unsigned char *buf, uint32_t *buff_len, uint32_t *num)
+{
+  uint8_t i, len, type;
+
+  buf = snmp_ber_decode_type(buf, buff_len, &type);
+
+  if(type != BER_DATA_TYPE_INTEGER) {
+    /*
+     * Sanity check
+     * Invalid type in buffer
+     */
+    return NULL;
+  }
+
+  buf = snmp_ber_decode_length(buf, buff_len, &len);
+
+  if(len > 4) {
+    /*
+     * Sanity check
+     * It will not fit in the uint32_t
+     */
+    return NULL;
+  }
+
+  *num = (uint32_t)(*buf++ & 0xFF);
+  (*buff_len)--;
+  for(i = 1; i < len; ++i) {
+    *num <<= 8;
+    *num |= (uint8_t)(*buf++ & 0xFF);
+    (*buff_len)--;
+  }
+
+  return buf;
+}
+/*---------------------------------------------------------------------------*/
+unsigned char *
+snmp_ber_decode_unsigned_integer(unsigned char *buf, uint32_t *buff_len, uint8_t expected_type, uint32_t *num)
+{
+  uint8_t i, len, type;
+
+  buf = snmp_ber_decode_type(buf, buff_len, &type);
+
+  if(type != expected_type) {
+    /*
+     * Sanity check
+     * Invalid type in buffer
+     */
+    return NULL;
+  }
+
+  buf = snmp_ber_decode_length(buf, buff_len, &len);
+
+  if(len > 4) {
+    /*
+     * Sanity check
+     * It will not fit in the uint32_t
+     */
+    return NULL;
+  }
+
+  *num = (uint32_t)(*buf++ & 0xFF);
+  (*buff_len)--;
+  for(i = 1; i < len; ++i) {
+    *num <<= 8;
+    *num |= (uint8_t)(*buf++ & 0xFF);
+    (*buff_len)--;
+  }
+
+  return buf;
+}
+/*---------------------------------------------------------------------------*/
+unsigned char *
+snmp_ber_decode_string_len_buffer(unsigned char *buf, uint32_t *buff_len, const char **str, uint32_t *length)
+{
+  uint8_t type;
+
+  buf = snmp_ber_decode_type(buf, buff_len, &type);
+
+  if(type != BER_DATA_TYPE_OCTET_STRING) {
+    /*
+     * Sanity check
+     * Invalid type in buffer
+     */
+    return NULL;
+  }
+
+  uint8_t i, length_bytes;
```

Comment:
Please move to the start of the function. We typically don't intermingle variable definitions with code (for historic reasons)

Reply:
This was a mistake. I also don't like this.